spark/src/main/scala/io/substrait/spark/ToSubstraitType.scala

```diff
@@ -42,11 +44,47 @@ private class ToSparkType
 
   override def visit(expr: Type.Str): DataType = StringType
 
+  override def visit(expr: Type.Binary): DataType = BinaryType
+
   override def visit(expr: Type.FixedChar): DataType = StringType
 
   override def visit(expr: Type.VarChar): DataType = StringType
 
   override def visit(expr: Type.Bool): DataType = BooleanType
+
+  override def visit(expr: Type.PrecisionTimestamp): DataType = {
+    if (expr.precision() != Util.MICROSECOND_PRECISION) {
+      throw new UnsupportedOperationException(
+        s"Unsupported precision for timestamp: ${expr.precision()}")
+    }
+    TimestampNTZType
+  }
+  override def visit(expr: Type.PrecisionTimestampTZ): DataType = {
+    if (expr.precision() != Util.MICROSECOND_PRECISION) {
+      throw new UnsupportedOperationException(
+        s"Unsupported precision for timestamp: ${expr.precision()}")
+    }
+    TimestampType
+  }
+
+  override def visit(expr: Type.IntervalDay): DataType = {
+    if (expr.precision() != Util.MICROSECOND_PRECISION) {
```

Comment:
To confirm, the Spark `DayTimeIntervalType` encodes time down to seconds, but can store fractional seconds.

From the Spark code
> Internally, values of day-time intervals are stored in Long values as amount of time in terms of microseconds that are calculated by the formula: -/+ (24*60*60 * DAY + 60*60 * HOUR + 60 * MINUTE + SECOND) * 1000000

So the default / supported precision is microseconds?

Reply:
Yea, Spark always stores the value as microseconds. 

It's possible this is unnecessarily strict, I checked that [on DataFusion we don't care](https://github.com/apache/datafusion/blob/6a3c0b0bce67553a4431b941d13fc995f310bee8/datafusion/substrait/src/logical_plan/consumer.rs#L1833) about the precision at all for the type conversion.. but I guess it doesn't hurt to be strict to begin with, we can relax it later if there's need.

Reply:
Let's start strict and see how we go.

> I checked that [on DataFusion we don't care](https://github.com/apache/datafusion/blob/6a3c0b0bce67553a4431b941d13fc995f310bee8/datafusion/substrait/src/logical_plan/consumer.rs#L1833) about the precision at all for the type conversion.

That's the case for now, but that may also change if we want to enforce semantics with Substrait strictly.

Reply:
Yea, I was thinking about that today - problem is I do need Spark -> DataFusion to play nicely together and that's tough if Spark insists strictly on one precision and DF on another 😅 but then what is the risk of things going wrong if the precision don't match, that I'm not sure about.. on literals it all makes sense, but for types themselves?